src/main/java/io/neonbee/data/DataVerticle.java

```diff
@@ -548,18 +552,28 @@ public Future<T> execute(DataQuery query, DataContext context) {
             // order, as the collection returned via requireData. This also favours the previous implementation of
             // requireData(), where any index of the requireData array corresponded with the indexes of the data array
             Map<DataRequest, AsyncResult<?>> requestResults = new LinkedHashMap<>();
+            Map<DataRequest, DataContext> receivedDataContextMap = new LinkedHashMap<>();
             return requireData(query, context).compose(requests -> {
                 // ignore the result of the require data composite future (otherwiseEmpty), the retrieve data method
                 // should decide if it needs to handle success or failure of any of the individual asynchronous results
-                return CompositeFuture.join(Optional.ofNullable(requests).map(Collection::stream).orElse(Stream.empty())
-                        .map(request -> requestResults.computeIfAbsent(request, mapRequest -> {
-                            Future<Object> future = requestData(vertx, request, context.copy());
-                            reportRequestDataMetrics(request, future);
-                            return future;
-                        })).map(Future.class::cast).collect(Collectors.toList())).otherwiseEmpty();
+                return CompositeFuture.join(
+                        Optional.ofNullable(requests).map(Collection::stream).orElse(Stream.empty()).map(request -> {
+                            // use one copy of DataContext for each request to avoid data clash
+                            DataContext requestContext = context.copy();
+                            receivedDataContextMap.put(request, requestContext);
```

Comment:
does it make sense to only already store the receivedData here, then we could maybe spare us the stream().map operation of the map in line 573-575 or why else is the whole map of the context required?

Reply:
I don't think so, we are keeping the reference to the DataContext before sending the requests via the CompositeFuture. At this time, we don't have reference to the receivedData, since they are not yet allocated at this time. The only reference we have is the DataContext wrapper.